blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
+				dag.reachabilityStore.setTreeNode(modifiedTreeNode)
+			}
+		}
+	}
+
 	return nil
 }
 
+func (dag *BlockDAG) updateReachabilityReindexRoot(newTreeNode *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	nextReindexRoot := dag.reachabilityReindexRoot
+	for {
+		candidateReindexRoot, modifiedNodes, found, err := dag.tryMovingReachabilityReindexRoot(nextReindexRoot, newTreeNode)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			break
+		}
+		for node := range modifiedNodes {
+			modifiedTreeNodes[node] = struct{}{}
+		}
+		nextReindexRoot = candidateReindexRoot
+	}
+
+	dag.reachabilityReindexRoot = nextReindexRoot
+	return modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tryMovingReachabilityReindexRoot(
+	reindexRoot *reachabilityTreeNode, newTreeNode *reachabilityTreeNode) (
+	newReindexRoot *reachabilityTreeNode, modifiedTreeNodes map[*reachabilityTreeNode]struct{}, found bool, err error) {
+
+	if !reindexRoot.isAncestorOf(newTreeNode) {
+		commonAncestor := reindexRoot.findCommonAncestor(newTreeNode)
+		return commonAncestor, nil, true, nil
+	}
+
+	chosenReindexRootChild, err := reindexRoot.findReachabilityTreeAncestorInChildren(newTreeNode)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	if newTreeNode.blockNode.blueScore-chosenReindexRootChild.blockNode.blueScore < reachabilityReindexWindow {
+		return nil, nil, false, nil
+	}
+	modifiedTreeNodes, err = dag.concentrateReachabilityTreeIntervalAroundReindexRootChild(reindexRoot, chosenReindexRootChild)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	return chosenReindexRootChild, modifiedTreeNodes, true, nil
+}
+
+// findReachabilityTreeAncestorInChildren finds the reachability tree child
+// of rtn that is the ancestor of node.
+func (rtn *reachabilityTreeNode) findReachabilityTreeAncestorInChildren(node *reachabilityTreeNode) (*reachabilityTreeNode, error) {
+
+	rootChildrenFutureCoveringSet := futureCoveringBlockSetFromReachabilityTreeNodes(rtn.children)
+	i := rootChildrenFutureCoveringSet.findIndex(&futureCoveringBlock{blockNode: node.blockNode, treeNode: node})
+	if i == 0 {
+		return nil, errors.Errorf("rtn is not an ancestor of node")
+	}
+
+	return rootChildrenFutureCoveringSet[i-1].treeNode, nil
+}
+
+func (dag *BlockDAG) concentrateReachabilityTreeIntervalAroundReindexRootChild(
+	reindexRoot *reachabilityTreeNode, chosenReindexRootChild *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	reindexRootChildNodesBeforeChosen, reindexRootChildNodesAfterChosen, err :=
+		reindexRoot.splitChildrenAroundChosenChild(chosenReindexRootChild)
+	if err != nil {
+		return nil, err
+	}
+
+	reindexRootChildNodesBeforeChosenSizesSum, modifiedNodes, err :=
+		dag.tightenReachabilityTreeIntervalsBeforeChosenReindexRootChild(reindexRoot, reindexRootChildNodesBeforeChosen)
```

Comment:
maybe tightenReachabilityTreeIntervalsBeforeReindexRoot**Chosen**Child

also I think TreeIntervals can be Intervals

Reply:
I think `ReachabilityTreeIntervals` can be `Intervals`, in all names, not only here.

The thing is, those functions can't be dag functions in that case.
However, from a quick check, it seems like we are never actually using the dag, only dag.reachabilityStore, so maybe we can scope all these functions to reachabilityStore? (We might want to rename it to reachabilityManager or something to be more general)